### PR TITLE
BUGFIX: AllowedEmail has wrong type of version column and add unit tests

### PR DESCRIPTION
**Pull request by caohhung**
Please review and test it carefully

### CODE REVIEW

File: DanceCat/Models.py [hunk 1]
```diff
@@ -70,7 +70,10 @@ def __init__(self, allowed_email):
             The email that will be allow to
             register a new user.
         """
-        self.email = allowed_email
+        if Helpers.is_valid_format_email(allowed_email):
+            self.email = allowed_email
+        else:
+            raise ValueError('Email %s is not well-known format')
```
Comment (scattm) [inline DanceCat/Models.py]: Email %s is not valid!

File: DanceCat/Models.py [hunk 2]
```diff
@@ -111,7 +114,10 @@ def __init__(self, user_email, user_password):
         :param user_email: User's email.
         :param user_password: User's password in clear text.
         """
-        self.email = user_email
+        if Helpers.is_valid_format_email(user_email):
+            self.email = user_email
+        else:
+            raise ValueError('Email %s is not well-known format' % user_email)
```
Comment (scattm) [inline DanceCat/Models.py]: Same as above

File: tests/test_models.py [hunk 3]
```diff
@@ -27,7 +27,6 @@ class TestUserModels(object):
     }
 
     def test_should_add_new_users(self, app):
-        """Test add_allowed_user command."""
```
Comment (scattm) [inline tests/test_models.py]: Should also remove this empty line

File: tests/test_dance_cat_console.py [hunk 4]
```diff
@@ -47,7 +47,7 @@ def test_add_allowed_user(app, user_email, capfd):
 
 def test_schedule_update(app, user_email):
     """Test schedule_update command."""
-    allowed_email = Console.Models.AllowedEmail(user_email)
+    allowed_email = Models.AllowedEmail(user_email)
     db.session.add(allowed_email)
     db.session.commit()
     assert allowed_email.email
```
Comment (scattm) [inline tests/test_dance_cat_console.py]: L71: Please change from 
`job = Models.Job(`
to
`job = Models.QueryDataJob(`
Thanks 🙏

File: tests/test_models.py [hunk 5]
```diff
@@ -121,3 +120,66 @@ def test_should_set_last_updated_valid_case(self, app):
         assert user.last_updated
         assert round(datetime_delta_last_updated.total_seconds()) == 0
         assert user.last_updated != previous_last_updated
+
+    def test_user_is_anonymous(self, app):
```
Comment (scattm) [inline tests/test_models.py]: You mean *is_active*?

File: tests/test_models.py [hunk 6]
```diff
@@ -121,3 +120,66 @@ def test_should_set_last_updated_valid_case(self, app):
...
+class TestAllowedEmailModels(object):
+    """ Unit tests for Models.AllowedEmail class. """
+
+    user_email = 'test@test.test'
+
+    def test_should_able_to_add(self, app, capfd):
+        allowed_email = Models.AllowedEmail(allowed_email=self.user_email)
+        db.session.add(allowed_email)
+        db.session.commit()
+        assert allowed_email.version == 1
```
Comment (scattm) [inline tests/test_models.py]: Should be
`assert allowed_email.version == Constants.MODEL_ALLOWED_EMAIL_VERSION`

Also, update L140 please 🐵

File: tests/test_models.py [hunk 7]
```diff
@@ -121,3 +120,66 @@ def test_should_set_last_updated_valid_case(self, app):
...
+        db.session.commit()
+        allowed_email_2 = Models.AllowedEmail(allowed_email=self.user_email)
+        db.session.add(allowed_email_2)
+        with pytest.raises(sqlalchemy_exc.IntegrityError):
+            db.session.commit()
+
+    def test_should_not_add_invalid_email_address(self, app):
+        with pytest.raises(ValueError):
+            allowed_email = Models.AllowedEmail(allowed_email='Invalid')
+
```
Comment (scattm) [inline tests/test_models.py]: Could you change this file to *test_model_users* or s.th relevant?

File: tests/test_models.py [hunk 8]
```diff
@@ -121,3 +120,66 @@ def test_should_set_last_updated_valid_case(self, app):
...
+        user = Models.User(**self.user_1)
+        db.session.add(user)
+        db.session.commit()
+        print (user)
+        out, err = capfd.readouterr()
+        assert out == '<User test@test.test - Id 1>\n'
+        assert not err
+
+
+class TestAllowedEmailModels(object):
```
Comment (scattm) [inline tests/test_models.py]: Should there be no *s* at the end?

**Comment by scattm** [comment]
Reviewed.

Thank Hung 👍

File: tests/test_models.py [hunk 9]
```diff
@@ -121,3 +120,66 @@ def test_should_set_last_updated_valid_case(self, app):
...
+        db.session.add(allowed_email)
+        db.session.commit()
+        allowed_email_2 = Models.AllowedEmail(allowed_email=self.user_email)
+        db.session.add(allowed_email_2)
+        with pytest.raises(sqlalchemy_exc.IntegrityError):
+            db.session.commit()
+
+    def test_should_not_add_invalid_email_address(self, app):
+        with pytest.raises(ValueError):
+            allowed_email = Models.AllowedEmail(allowed_email='Invalid')
```
Comment (scattm) [inline tests/test_models.py]: What will happened if we put valid email here?
Also, try some:
- user@mail
- usermail@
- @usermail
- user@mail.
- user@.mail
- user.mail@domain
- user.@mail
- .user@mail

File: DanceCat/Models.py (hunk 1 above)
Comment (scattm) [inline DanceCat/Models.py]: 'Email {0} is not valid!'.format(allowed_email)

File: tests/test_models.py (hunk 7 above)
Comment (caohhung) [inline tests/test_models.py]: No need, I separated all tests for each class into test-classes

Comment (scattm) [inline tests/test_models.py]: I just want to keep the file not too long, plus won't conflict to other test-classes on the same file if two people work on 2 or more classes.

File: tests/test_models.py (hunk 9 above)
Comment (caohhung) [inline tests/test_models.py]: All cases above are caught by 're.search('^.*?@[\w-]+\.[\w-]+$', email, flags=re.IGNORECASE)' except 'user-mail@domain.com', is valid email address

File: tests/test_models.py (hunk 7 above)
Comment (caohhung) [inline tests/test_models.py]: It's ok for a long test file. Conflict always happen and we could handle it easily in TestClass, don't worry

**Comment by caohhung** [comment]
Please let me know when you done your review, I'll squash 2 commits

File: tests/test_models.py (hunk 9 above)
Comment (scattm) [inline tests/test_models.py]: Well, I think we should add to Helper to check that function. I will handle it later.

Comment (caohhung) [inline tests/test_models.py]: you mean unittest for this function?

Comment (scattm) [inline tests/test_models.py]: Yes in Helpers, I will cover that.

**Comment by scattm** [comment]
Done, please squash it

**Pull request merged**